File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
	static List<int[]> list;
	static Map<String, Integer> map;
    public int[] solution(String[] genres, int[] plays) {
        int[] answer = {};
    	// 장르별로 가장 많이 재생된 노래를 두 개씩 모아 베스트 앨범 출시
    	// 노래는 고유 번호로 구분
    	map = new HashMap<>();
    	for(int i = 0; i < plays.length; i++)
    		map.put(genres[i], map.getOrDefault(genres[i], 0)+plays[i]);
    	List<String> keySet = new ArrayList<>(map.keySet());
    	// key값가지고 value값을 기준으로 정렬하는 방법..(이거를 어떻게 외우냐고)
    	keySet.sort((o1,o2) -> map.get(o2).compareTo(map.get(o1)));
    	List<Integer> result_list = new ArrayList<>();
    	for(int i = 0; i < keySet.size(); i++) {
    		list = new ArrayList<>();
    		for(int j = 0; j < plays.length; j++) {
    			if(keySet.get(i).equals(genres[j])) {
    				list.add(new int[] {plays[j], j});
    			}
    		}
    		// int[] list 정렬하는 방법 ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 환장하겠다
        	Collections.sort(list, new Comparator<int []>() {
        		@Override
        		public int compare(int[] a, int[] b) {
        			return a[0] != b[0] ? b[0] - a[0] : a[1] - b[1];
        		}
        	});
        	for(int j = 0; j < list.size(); j++) {
        		if(j >= 2)
        			break;
        		result_list.add(list.get(j)[1]);
        	}
    	}
    	int [] result = new int[result_list.size()];
    	for(int i = 0; i < result_list.size(); i++) {
    		result[i] = result_list.get(i);
    	}
    	answer = result;
        return answer;
    }
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
from collections import defaultdict
def solution(genres, plays):
    answer = []
    music_dict = defaultdict(list)
    total_play = defaultdict(int)
    for i in range(len(genres)):
        music_dict[genres[i]].append((i, plays[i]))
        total_play[genres[i]] += plays[i]
    # 복잡한 리스트 dict 키 정렬하는 방법
    for key, items in music_dict.items():
        music_dict[key] = sorted(items, key=lambda item: (-item[1], item[0]))
    total_play = sorted(total_play.items(), key=lambda item: -item[1])
    for key, items in total_play:
        if key in music_dict:
            music_list = music_dict[key][:2]
            for song in music_list:
                answer.append(song[0])
    return answer